File: updater/platform.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

// Platform represents the detected platform information
type Platform struct {
	OS   string // windows, darwin, linux
	Arch string // amd64, arm64, 386
}

// DetectPlatform detects the current operating system and architecture
func DetectPlatform() *Platform {
	return &Platform{
		OS:   runtime.GOOS,
		Arch: normalizeArch(runtime.GOARCH),
	}
}

// normalizeArch normalizes architecture names
func normalizeArch(arch string) string {
	switch arch {
	case "amd64", "x86_64":
		return "amd64"
	case "arm64", "aarch64":
		return "arm64"
	case "386", "x86":
		return "386"
	default:
		return arch
	}
}

// IsWindows returns true if the platform is Windows
func (p *Platform) IsWindows() bool {
	return p.OS == "windows"
}

// IsMacOS returns true if the platform is macOS
func (p *Platform) IsMacOS() bool {
	return p.OS == "darwin"
}

// IsLinux returns true if the platform is Linux
func (p *Platform) IsLinux() bool {
	return p.OS == "linux"
}

// GetInstallDir returns the default installation directory for the platform
func (p *Platform) GetInstallDir() string {
	switch p.OS {
	case "windows":
		return `C:\Program Files\OpenClaw`
	case "darwin":
		return "/usr/local/bin"
	case "linux":
		return "/usr/local/bin"
	default:
		return "/usr/local/bin"
	}
}

// GetConfigDir returns the configuration directory for the platform
func (p *Platform) GetConfigDir() string {
	switch p.OS {
	case "windows":
		return `C:\ProgramData\OpenClaw`
	case "darwin":
		return "/etc/openclaw"
	case "linux":
		return "/etc/openclaw"
	default:
		return "/etc/openclaw"
	}
}

// GetConfigPath returns the default configuration file path
func (p *Platform) GetConfigPath() string {
	return filepath.Join(p.GetConfigDir(), "openclaw.json")
}

// GetBackupDir returns the default backup directory for the platform
func (p *Platform) GetBackupDir() string {
	switch p.OS {
	case "windows":
		return `C:\ProgramData\OpenClaw\backups`
	case "darwin":
		return "/var/lib/openclaw/backups"
	case "linux":
		return "/var/lib/openclaw/backups"
	default:
		return "/var/lib/openclaw/backups"
	}
}

// GetDataDir returns the default data directory for the platform
func (p *Platform) GetDataDir() string {
	switch p.OS {
	case "windows":
		return `C:\ProgramData\OpenClaw\data`
	case "darwin":
		return "/var/lib/openclaw"
	case "linux":
		return "/var/lib/openclaw"
	default:
		return "/var/lib/openclaw"
	}
}

// GetLogDir returns the default log directory for the platform
func (p *Platform) GetLogDir() string {
	switch p.OS {
	case "windows":
		return `C:\ProgramData\OpenClaw\logs`
	case "darwin":
		return "/var/log/openclaw"
	case "linux":
		return "/var/log/openclaw"
	default:
		return "/var/log/openclaw"
	}
}

// GetBinaryName returns the binary name with extension for the platform
func (p *Platform) GetBinaryName(name string) string {
	if p.IsWindows() {
		return name + ".exe"
	}
	return name
}

// String returns a string representation of the platform
func (p *Platform) String() string {
	return fmt.Sprintf("%s/%s", p.OS, p.Arch)
}

// GetPackageExtension returns the package extension for the platform
func (p *Platform) GetPackageExtension() string {
	if p.IsWindows() {
		return ".zip"
	}
	return ".tar.gz"
}

// GetPackageName generates the package name for a component
func (p *Platform) GetPackageName(component string) string {
	ext := p.GetPackageExtension()
	return fmt.Sprintf("%s-%s-%s%s", component, p.OS, p.Arch, ext)
}

// EnsureDirectories creates all necessary directories
func (p *Platform) EnsureDirectories() error {
	dirs := []string{
		p.GetInstallDir(),
		p.GetConfigDir(),
		p.GetDataDir(),
		p.GetLogDir(),
		p.GetBackupDir(),
	}

	for _, dir := range dirs {
		if err := os.MkdirAll(dir, 0755); err != nil {
			return fmt.Errorf("failed to create directory %s: %w", dir, err)
		}
	}

	return nil
}

File: packages/shared/utils/platform.ts
import { platform, arch } from 'os'
import { join } from 'path'
import type { PlatformInfo, PlatformPaths, OS, Arch } from '../types'

export function detectPlatform(): PlatformInfo {
  const os = platform()
  const cpuArch = arch()

  const normalizedOS: OS = os === 'win32' ? 'windows' : os as OS
  const normalizedArch: Arch = cpuArch === 'x64' ? 'amd64' : 'arm64'

  return {
    os: normalizedOS,
    arch: normalizedArch,
    isWindows: os === 'win32',
    isMacOS: os === 'darwin',
    isLinux: os === 'linux',
  }
}

export function getPlatformPaths(info?: PlatformInfo): PlatformPaths {
  const p = info ?? detectPlatform()

  if (p.isWindows) {
    const appData = process.env.APPDATA || join(process.env.USERPROFILE || '', 'AppData', 'Roaming')
    return {
      installDir: join(process.env.PROGRAMFILES || 'C:\\Program Files', 'OpenClaw'),
      configDir: join(appData, 'OpenClaw'),
      configFile: join(appData, 'OpenClaw', 'config.json'),
      logDir: join(appData, 'OpenClaw', 'logs'),
    }
  }

  if (p.isMacOS) {
    const home = process.env.HOME || '/Users'
    return {
      installDir: '/usr/local/bin',
      configDir: join(home, 'Library', 'Application Support', 'OpenClaw'),
      configFile: join(home, 'Library', 'Application Support', 'OpenClaw', 'config.json'),
      logDir: join(home, 'Library', 'Logs', 'OpenClaw'),
    }
  }

  // Linux
  const home = process.env.HOME || '/home'
  return {
    installDir: '/usr/local/bin',
    configDir: join(home, '.config', 'openclaw'),
    configFile: join(home, '.config', 'openclaw', 'config.json'),
    logDir: join(home, '.local', 'share', 'openclaw', 'logs'),
  }
}
